Alerting: Improve notification policies created during migration

JacobsonMT opened the pull request:
Previously, migrated legacy alerts were connected to notification policies through a `rule_uid` label in a 1:1 fashion. While this correctly mimicked pre-migration routing, it didn't create a notification policy structure that is easy to view/modify. In addition,
having one policy per migrated alert is, in some ways, counter to the recommended approach of Unified Alerting.

This change replaces `rule_uid`-based migrated notification policies with a private label called `__contacts__`. This label stores a list of double quoted strings containing the names of all contact points an AlertRule should route to (based on legacy notification channels). Then, one notification policy is created per contact point matching AlertRules via regex on this label.

**NOTE**: In order to simplify the created regex, we sanitize any legacy notification channel names that contain a double quote, replacing it with an underscore.

The result is a simpler, clearer, and easier to modify notification policy structure, with the added benefit that you can see which contact points an AlertRule is being routed to from the AlertRule creation page.

Fixes https://github.com/grafana/grafana/issues/56582

### Specific Example:

#### Legacy Setup

| Alert | Channels |
| ------------- | ------------- |
| Alert1 | email1, webhook |
| Alert2 | email2, webhook |
| Alert3 | webhook |
| Alert4 | email2 |
| Alert5 | email2 |

#### Previous Migrated Setup
5 Notification Policies: 
- One per alert, matched based on the alert's `rule_uid`
- Two (Alert1, Alert2) will have nested policies so that they send to multiple contact points.

![Screenshot from 2022-07-11 18-22-37](https://user-images.githubusercontent.com/8484471/178368649-90741579-4ec6-4617-af46-30beb5f02b5a.png)

#### New Migrated Setup
3 Notification Policies: 
- One per contact point, matched based on the `__contacts__` label regex, ex: `__contacts__=~.*"webhook".*`
- No nested policies.
- Policies match a collection of alerts instead of a single alert.
- Can add/remove which contact points are sent to on the alert rule page without changing policies by adding to the `__contacts__` label.
- Easy to see the purpose of each policy

![Screenshot from 2022-10-17 20-12-55](https://user-images.githubusercontent.com/8484471/196306426-14595ebc-127a-45e7-89f5-562d4088c348.png)

Alert1 Labels:
![Screenshot from 2022-10-17 20-13-57](https://user-images.githubusercontent.com/8484471/196306497-5d931f9a-45e6-437b-826e-8d997e7e369a.png)

JacobsonMT committed:
Alerting: Improve notification policies created during migration

Previously, migrated legacy alerts were connected to notification policies through
a `rule_uid` label in a 1:1 fashion. While this correctly mimicked pre-migration routing,
it didn't create a notification policy structure that is easy to view/modify. In addition,
having one policy per migrated alert is, in some ways, counter to the recommended approach of
Unified Alerting.

This change replaces `rule_uid`-based migrated notification policies with a private
label called `__contacts__`. This label stores a json-array string containing the names of
all contact points an AlertRule should route to (based on legacy notification channels). Then,
one notification policy is created per contact point matching AlertRules via regex on this
`__contacts__` label.

The result is a simpler, clearer, and easier to modify notification policy structure, with the
added benefit that you can see which contact points an AlertRule is being routed to from the
AlertRule creation page.

JacobsonMT committed:
Fix linting

joeblubaugh commented:
I like this idea. How do we set the "consumption" policy in this case? ("Continue matching subsequent sibling nodes")? I *think* they all would need to be set with that to "true" for this approach to work.

JacobsonMT commented:
> I like this idea. How do we set the "consumption" policy in this case? ("Continue matching subsequent sibling nodes")? I _think_ they all would need to be set with that to "true" for this approach to work.

Correct, each route is set to `continue=true` so that all siblings can match independently. See https://github.com/grafana/grafana/pull/52071/files#diff-3d6f33c060b6bf9fcfcb6c64e52c81198f5793d5ee389dd81395633cd8ea50f8R278.

JacobsonMT commented:
> An interesting question came up, that I want to relay here. Should we have a single label on the alertRule (`__contacts__`, `migrated_contacts`, ... = `["webhook","email1"]`) that contains all contacts and is matched using regex or should we have multiple labels (`webhook=true`, `email1=true`) that match without needing regex?

Alernative would look something like this (given example setup in description):

![Screenshot from 2022-07-15 19-39-06](https://user-images.githubusercontent.com/8484471/179324964-ebaec7d0-a9a6-4a8b-94b6-6742a6592153.png)

![Screenshot from 2022-07-15 19-39-50](https://user-images.githubusercontent.com/8484471/179324968-d5a28ab0-78ba-4f2e-996f-bd955691b160.png)

yuri-tceretian commented:
I like the idea. Do we control the name of the contact points? If yes, I think the single label with regex in notification policies would be a good solution.

JacobsonMT commented:
> I like the idea. Do we control the name of the contact points? If yes, I think the single label with regex in notification policies would be a good solution.

Yes, we control the contact point names. They are migrated as is from the legacy notification channel names which are unique (https://github.com/grafana/grafana/blob/643d2bc890301526de36518a3c2c8c3ab0d718bb/pkg/services/alerting/store_notification.go#L308).

yuri-tceretian reviewed (approved):
LGTM! 👍🏻

JacobsonMT committed:
Merge branch 'main' into jacobsonmt/improve-notification-policies-after-migration

JacobsonMT committed:
Merge branch 'main' into jacobsonmt/improve-notification-policies-after-migration

JacobsonMT committed:
Fix route matching when notification channel names contain reserved regex characters

Modifies the migration to create ObjectMatchers instead of Matchers.

grobinson-grafana commented on pkg/services/sqlstore/migrations/ualert/channel.go:
```diff
@@ -115,19 +114,20 @@ func (m *migration) setupAlertmanagerConfigs(rulesPerOrg map[int64]map[*alertRul
-	return string(jsonStr), nil
+// quote will surround the given string in double quotes and escape any existing interior double quotes.
+func quote(s string) string {
+	return `"` + strings.ReplaceAll(s, `"`, `\"`) + `"`
```
You might need to escape existing escaped quotes?

grobinson-grafana replied:
I'm not 100% sure on this, but if we have `\"` then we will get `\\"` when what we want I think is `\\\"`.

JacobsonMT replied:
You're right but extra escaping won't help either. Now that you point this out, I realize that my original intention for escaping quotes here doesn't accomplish what I had hoped.

Basically, I was attempting to use the outside quotes to separate contact names to make sure that partial names can't accidentally be matched by the policy regex `.*"<name>".*`. So, I escaped interior quotes so they wouldn't match the boundary. This doesn't work though, for example we could have two (rather evil) contact points:

- `hello\`
- `hello"world`

The regex for `hello\` would be `.*"hello\\".*` which would also incorrectly match `hello"world`'s contact label which would be `"hello\"world"`. [regex101 demonstration](https://regex101.com/r/3xOxye/1).

I'm sure there's a way to make the regex more robust using some lookahead or alternation but it won't be pretty... might be "good enough" in this state. Alternatively we can go the non-regex route and create multiple labels on each alert rule (one per contact).

JacobsonMT replied:
I think that the regex method is too brittle and too much hassle for the benefit. I'm going to push a change that uses multiple labels shortly and edit the description/screenshots.

JacobsonMT replied:
After some internal discussion, we decided to go back to the regex approach but sanitize any double quotes internal to the contact name on migration and replace them with `_`

JacobsonMT committed:
Revert "Fix route matching when notification channel names contain reserved regex characters"

This reverts commit 2862346be8b935298421c293c0825167dbbcb719.

JacobsonMT committed:
Change strategy to create multiple labels and simpler Matchers

Regex based matching on a list of contact point names was too brittle for the benefits it brings.
Instead I create a much simpler Matcher that just checks for existence of a label and add one label to each alert rule per contact point.

JacobsonMT committed:
Update migration_test.go for new strategy

JacobsonMT committed:
Merge branch 'main' into jacobsonmt/improve-notification-policies-after-migration

JacobsonMT committed:
Revert "Update migration_test.go for new strategy"

This reverts commit c3d9bb33b541f0e8d89bb3aee7219eacde9a4572.

JacobsonMT committed:
Revert "Change strategy to create multiple labels and simpler Matchers"

This reverts commit 4729858a7970cea78fc40e65aa756035b9b6d38b.

JacobsonMT committed:
Revert "Revert "Fix route matching when notification channel names contain reserved regex characters""

This reverts commit b96f5028cda10ce01c6b2bbb384f868214d88aed.

JacobsonMT committed:
Sanitize double quotes to underscore in migrated contact name

This prevents an edge case that allows for partial name matches in the Route Matcher.
Since we now sanitize double quotes we must also check for collisions, so in
case of a collision we append a short hash of the original name to the end.

joeblubaugh commented on pkg/services/sqlstore/migrations/ualert/channel.go:
```diff
@@ -114,7 +115,7 @@ func (m *migration) setupAlertmanagerConfigs(rulesPerOrg map[int64]map[*alertRul
 
-// contactListToString creates a sorted string representation of a given map (set) of receiver names. Each name will be comma-separated and double-quoted. Double quotes inside a name are escaped.
+// contactListToString creates a sorted string representation of a given map (set) of receiver names. Each name will be comma-separated and double-quoted. Names should not contain double quotes.
 func contactListToString(m map[string]interface{}) string {
```
Why is this restriction being relaxed?

(for my edification, not questioning the reason for the change)

joeblubaugh replied:
(I *think* it's because we already replace " with _)
